Fix .NET Core 3.X running WPF/WinForms

Sewer56 opened the pull request:
## Problem
In .NET Core 3.X WPF/WinForms applications the entry assembly is not the file used to launch the program. This causes `Onova.Updater` will try to launch a DLL on restart.

## Issue (Behind the Scenes)
When running a .NET Core 3.X WPF or WinForms application, the actual executing code of the main module lies in a different assembly than the file used to start up the program.

i.e. For project `MyProject`, .NET Core 3.X will export:
- MyProject.dll
- MyProject.exe

In this case, despite `MyProject.exe` being used to start the application, `MyProject.dll` is actually the assembly containing all of the executing code. `MyProject.exe` is only a bootstrapper to launch the MyProject DLL with .NET Core, which is considered the Entry Assembly (Assembly.GetEntryAssembly()).

## Pull Request

The following commit modifies the extension when creating `AssemblyMetadata` from the entry assembly. If the current runtime is .NET Core and the file is a DLL, the path extension is changed to `.exe`.

As far as I'm aware, there isn't any way in which you could make the EXE and DLL names different and I've searched the web a bunch. At least we're safe for now but it still feels like a hacky solution.

Tyrrrz commented:
Intesting. But can we just simply make a reasonable assumption that if the entry assembly is a .dll then we need to replace it with .exe? It won't work on .NET Core projects that are not self-contained (i.e. that need to be run with `dotnet app.dll`), but we don't really have a solution in place to deal with those.

Sewer56 commented:
> Intesting. But can we just simply make a reasonable assumption that if the entry assembly is a .dll then we need to replace it with .exe? That way we won't need to check for target framework.
> 
> It won't work on .NET Core projects that are not self-contained (i.e. that need to be run with `dotnet app.dll`), but we don't really have a solution in place to deal with those.

That assumption sounds perfectly reasonable in all honesty. I added the framework check as a guard just to be absolutely sure.

That said, since I've done this PR, one thing came to mind. Throwing an exception with this message might not be the greatest idea for said not self-contained .NET Core Console projects and/or projects which don't want to reboot. It could maybe break some existing Core applications which don't have the reboot flag enabled. 

I think therefore the best choice of option would be to provide an example, and refer to it in the error message.

-----------
Handling reboots with .NET Core Console Applications.

A. Make a batch script (`MyProject.bat`)
```batch
dotnet MyProject.dll
```

B. To UpdateManager, pass custom AssemblyMetadata.
```csharp
var entryAssembly = Assembly.GetEntryAssembly();
var name        = entryAssembly.GetName().Name;
var version     = entryAssembly.GetName().Version;
var filePath    = Path.GetDirectoryName(entryAssembly.Location) + "\\MyProject.bat";

var metadata = new AssemblyMetadata(name, version, filePath);
using (var manager = new UpdateManager(metadata, ...))
...
```

-----------
I changed the PR to include such an example in the readme, and adjusted the error message.
Also made a change to explicitly ensure the working directory of the file to be executed is that file's directory. I've seen many programs depend on a given working directory.

Tyrrrz commented:
There's one place in the updater that implicitly requires that the provided `updateeFilePath` is actually loaded into memory during app's lifetime.

https://github.com/Tyrrrz/Onova/blob/49af8110698baaf84f9d8fc5b3f50dbd0bbd939e/Onova.Updater/Program.cs#L29-L32

Here it basically attempts to open a write stream on the executable in a loop. This is the best implementation I came up with so far for making sure the application has exited before Onova starts overwriting its files. `Process.GetProcesses` has caused a lot of issues in the past.

If a .bat file is supplied instead, this will give a false positive.

Sewer56 commented:
> There's one place in the updater that implicitly requires that the provided `updateeFilePath` is actually loaded into memory during app's lifetime.
> 
> https://github.com/Tyrrrz/Onova/blob/49af8110698baaf84f9d8fc5b3f50dbd0bbd939e/Onova.Updater/Program.cs#L29-L32
> 
> Here it basically attempts to open a write stream on the executable in a loop. This is the best implementation I came up with so far for making sure the application has exited before Onova starts overwriting its files. `Process.GetProcesses` has caused a lot of issues in the past.
> 
> If a .bat file is supplied instead, this will give a false positive.

It seems you're right, I did a mistake and automatically assumed that the command prompt would hold a handle to the batch file instead of reading it all at once. That was my bad.

It seems I'll have to re-evaluate possible options for console applications. I'll undo the commit that suggests the readme for now so the actual workaround for Core 3 WPF/Forms can be merged. Possibly I might contribute something for Core console application restarting another day.

Ideally the updater would just get another parameter for instead launching an alternative executable, settable from `UpdateManager`'s `LaunchUpdater` function, that sounds like the most straightforward implementation. I'd personally leave it until someone opens an issue for it though.

Sewer56 commented:
The solution to the problem above is rather simple, just check if every file is writable (no application holds handle to it) before copying.

https://github.com/Sewer56/Onova/blob/bd96e247831eae39ba2f94e4e8f68a7740fcc0a4/Onova.Updater/Internal/DirectoryEx.cs#L16-L27

This should provide a simple solution to the problem above.

Tyrrrz commented:
Good idea. But we probably should check that all files are writable before we write to any one file. Otherwise we risk overwriting some of the files that were not locked at the time, but are required by the running application. E.g. some resource files like images or text files may not be locked at all times but are still vital to the correct workflow of the application.

Tyrrrz commented:
Ideally, Onova.Updater should try to acquire locks for ALL files in the updatee directory and then proceed with the overwrite.

Tyrrrz commented:
I think the easiest, minimal effort solution to the original problem is to leave Onova as it is, but extend Onova.Updater so that it has some additional logic if the given file path is a .dll, i.e. try to find an .exe or launch via `dotnet`.

Tyrrrz commented:
So what do you think about that?

Sewer56 commented:
Life got me moving and I kept steamrolling my own projects since last week.

That said, that solution does also sound like it would work very well, could implement that and see how well it goes. 😉

Tyrrrz commented:
Cool. If you don't have time, I can just do it myself.

Sewer56 commented:
> Cool. If you don't have time, I can just do it myself.

Works well for me at the current time, please do.

I'm currently porting a few projects (mods) between two of my other projects (mod loaders), and originally forked Onova to extend it a bit for my own use case as well as helping fix a few minor things on the official repo.

**Extra note:**
Not sure if you're interested but my fork also has an extra branch where I replaced `Newtonsoft.Json` with the new faster official `System.Text.Json`. The new Json parser is part of .NET Core 3.0 and for future projects it will save an extra external dependency. 

https://github.com/Sewer56/Onova/commit/e0b446c9f50226f53eaf35cbba4959ee101057cd

Tyrrrz commented:
Done.

>Not sure if you're interested but my fork also has an extra branch where I replaced Newtonsoft.Json with the new faster official System.Text.Json. The new Json parser is part of .NET Core 3.0 and for future projects it will save an extra external dependency.

Very much interested, however I'm planning to wait until .NET Core 3.0 is finally released before migrating.
